src/ray/raylet/reconstruction_policy.cc

```diff
@@ -4,8 +4,146 @@ namespace ray {
+    int64_t initial_reconstruction_timeout_ms, const ClientID &client_id,
+    gcs::PubsubInterface<TaskID> &task_lease_pubsub,
+    std::shared_ptr<ObjectDirectoryInterface> object_directory)
+    : io_service_(io_service),
+      reconstruction_handler_(reconstruction_handler),
+      initial_reconstruction_timeout_ms_(initial_reconstruction_timeout_ms),
+      client_id_(client_id),
+      task_lease_pubsub_(task_lease_pubsub),
+      object_directory_(std::move(object_directory)) {}
+
+void ReconstructionPolicy::SetTaskTimeout(
+    std::unordered_map<TaskID, ReconstructionTask>::iterator task_it,
+    int64_t timeout_ms) {
+  task_it->second.expires_at = current_sys_time_ms() + timeout_ms;
+  auto timeout = boost::posix_time::milliseconds(timeout_ms);
+  task_it->second.reconstruction_timer->expires_from_now(timeout);
+  const TaskID task_id = task_it->first;
+  task_it->second.reconstruction_timer->async_wait(
+      [this, task_id](const boost::system::error_code &error) {
+        if (!error) {
+          auto it = listening_tasks_.find(task_id);
+          RAY_CHECK(it != listening_tasks_.end());
+
+          if (it->second.subscribed) {
+            // We did not receive a task lease notification within the lease period.
+            // The lease is expired, so attempt to reconstruct the task.
+            HandleTaskLeaseExpired(task_id);
+          } else {
+            // This task is still required, so subscribe to task lease notifications.
+            // Reconstruction will be triggered if the current task lease expires, or
+            // if no one has acquired the task lease.
+            task_lease_pubsub_.RequestNotifications(JobID::nil(), task_id, client_id_);
+            it->second.subscribed = true;
+          }
+        }
+      });
+}
+
+void ReconstructionPolicy::AttemptReconstruction(const TaskID &task_id,
+                                                 const ObjectID &object_id,
+                                                 int reconstruction_attempt) {
+  // If we are no longer listening for objects created this task, give up.
+  auto it = listening_tasks_.find(task_id);
+  if (it == listening_tasks_.end()) {
+    return;
+  }
+
+  if (it->second.created_objects.count(object_id) == 0) {
+    return;
+  }
+
+  // If we're already past this reconstruction attempt, give up.
+  if (reconstruction_attempt != it->second.reconstruction_attempt) {
+    return;
+  }
+  // Increment the number of times reconstruction has been attempted.
+  it->second.reconstruction_attempt++;
+
+  // Reset the timer to wait for task lease notifications again. NOTE(swang):
+  // The timer should already be set here, but we extend it to give some time
+  // for the reconstructed task to propagate notifications.
+  SetTaskTimeout(it, initial_reconstruction_timeout_ms_);
+  // TODO(swang): Suppress simultaneous attempts to reconstruct the task using
+  // the task reconstruction log.
+  reconstruction_handler_(task_id);
```

Comment:
Currently this will just resubmit the task, so the current behavior as of this PR is no suppression, right?

Reply:
Yeah, actually the current behavior of the PR is that the raylet will exit since `ResubmitTask` throws an error. I can change that to printing a warning.